[tplinksmarthome] Added support for HS210 and HS220, and some improvements

Hilbrand opened the pull request:
This pull request adds/improves the following to the tplink smart home binding:
* Added support for HS210 and HS220 (See https://community.openhab.org/t/tp-link-dimmer-switch-binding/42192)
* Improved support for TP-Link range extenders.
* Added category tags to thing types.

martinvw commented on addons/binding/org.openhab.binding.tplinksmarthome.test/src/test/java/org/openhab/binding/tplinksmarthome/handler/SmartHomeHandlerTest.java:
```diff
@@ -99,16 +99,31 @@ public void testInitializeShouldCallTheCallback() throws InterruptedException {
-    public void testHhandleCommandRefreshType() {
+    public void testHandleCommandRefreshType() {
+        handler.initialize();
+        assertHandleCommandRefreshType(-53);
```
Should this also be `expectedRssi`

Hilbrand replied:
I'm not sure what you mean with this comment about `expectedRssi`?

martinvw commented on addons/binding/org.openhab.binding.tplinksmarthome.test/src/test/java/org/openhab/binding/tplinksmarthome/internal/TPLinkSmartHomeDiscoveryServiceTest.java:
```diff
@@ -35,8 +40,12 @@
 public class TPLinkSmartHomeDiscoveryServiceTest {
 
+    private static final List<Object[]> TESTS = Arrays.asList(
+            new Object[][] { { "bulb_get_sysinfo_response_on", 11 }, { "rangeextender_get_sysinfo_response", 11 } });
```
Does the var-args way of calling `Arrays.asList` not work here?

Hilbrand replied:
Yes, but since it's a list of lists, it would be: ` Arrays.asList(new Object[] { "bulb_get_sysinfo_response_on", 11 }, new Object[]  { "rangeextender_get_sysinfo_response", 11 })`. I'm not sure if that would be an improvement.

martinvw commented on addons/binding/org.openhab.binding.tplinksmarthome/ESH-INF/thing/LB100.xml:
```diff
@@ -6,9 +6,10 @@
 
 		<channels>
-			<channel id="brightness" typeId="brightness" />
+	<channel id="brightness" typeId="brightness" />
```
Please keep the tabs intact

Hilbrand replied:
Ahhh. I keep messing up the tabs in the xml files... There were some other as well...

martinvw commented on addons/binding/org.openhab.binding.tplinksmarthome/src/main/java/org/openhab/binding/tplinksmarthome/handler/SmartHomeHandler.java:
```diff
@@ -90,7 +89,9 @@ public void initialize() {
         cache = new ExpiringCache<DeviceState>(TimeUnit.SECONDS.toMillis(configuration.refresh), this::refreshCache);
         updateStatus(ThingStatus.UNKNOWN);
-        startAutomaticRefresh(configuration);
+        if (configuration.refresh > 0) {
```
It seems to have a min value of 1 in the XML defs, does this make sense? Or is this test only?

Hilbrand replied:
Correct it for testing only. I've added an extra comment above the line to explain this.

Hilbrand commented:
@martinvw You added release `After 2.3`. But isn't everything that is not 2.3 after 2.3? Or did you intend this milestone to be pr's that are (almost) ready to be merged? Maybe name it 2.4 then? (While theoretical it could be pr's with milestone `After 2.3` could possible not being added in 2.4 either for some reason, so `After 2.3` is scientifically more sound I concur :smile:.

martinvw commented:
> Or did you intend this milestone to be pr's that are (almost) ready to be merged?

Correct, and I read somewhere that is going to be called 2.3-M1, so what does come after that M2? 2.4? I was planning to rename or remove it later. But it was meant as signal to others not to merge it in the code freeze :-)

martinvw commented:
```
[FATAL] Non-resolvable parent POM for org.openhab.binding:org.openhab.binding.tplinksmarthome:[unknown-version]: Could not find artifact org.openhab.binding:pom:pom:2.4.0-SNAPSHOT in sonatype-snapshots (https://oss.sonatype.org/content/repositories/snapshots/) and 'parent.relativePath' points at wrong local POM @ line 7, column 10
[FATAL] Non-resolvable parent POM for org.openhab.binding:org.openhab.binding.tplinksmarthome.test:[unknown-version]: Failure to find org.openhab.binding:pom:pom:2.4.0-SNAPSHOT in https://oss.sonatype.org/content/repositories/snapshots/ was cached in the local repository, resolution will not be reattempted until the update interval of sonatype-snapshots has elapsed or updates are forced and 'parent.relativePath' points at wrong local POM @ line 7, column 10
 @ 
```

martinvw commented:
Could you please rebase? @Hilbrand

Hilbrand commented:
Ok I was trying to be to clever by updating to the newer version before it was being released :smile: ... by creating a commit with updated version... removed and rebased....

martinvw commented:
Thanks!

Hilbrand commented:
@martinvw travis build failed (and on others too) Is that something that shouldn't be fixed? Something with versions?

martinvw commented:
Indeed but this should solve it

https://github.com/openhab/openhab2-addons/pull/3602
